File: Ruby/lib/civitas-ruby/controlador.rb
require_relative 'operacion_inmobiliaria'

module Civitas
  class Controlador
    
    def initialize(juego, vista)
      @juego = juego
      @vista = vista
    end

    def juega
      @vista.setCivitasJuego(@juego)
      operacion = nil
      while !@juego.final_del_juego
        #puts "Cuánto estás ciclando ami"
        @vista.actualizarVista
        @vista.pausa
        #puts "A = #{operacion}"
        @vista.mostrarSiguienteOperacion(operacion = @juego.siguiente_paso)
        #puts "B = #{operacion}"
        if operacion != OperacionesJuego::PASAR_TURNO
          @vista.mostrarEventos
        end
        fin = @juego.final_del_juego
        if !fin
          case operacion
          when OperacionesJuego::COMPRAR
            seleccion = @vista.comprar
            @juego.comprar if @vista.lista_respuestas[seleccion] == Respuestas::SI
            operacion = @juego.siguiente_paso_completado(operacion)
          when OperacionesJuego::GESTIONAR
            @vista.gestionar
            gest = @vista.i_gestion
            ip = @vista.i_propiedad
            op_inmobiliaria = OperacionInmobiliaria.new(gest, ip)
            case @vista.lista_gestiones_inmobiliarias[gest]
            when GestionesInmobiliarias::CANCELAR_HIPOTECA
              @juego.cancelar_hipoteca(ip)
            when GestionesInmobiliarias::CONSTRUIR_CASA
              @juego.construir_casa(ip)
            when GestionesInmobiliarias::CONSTRUIR_HOTEL
              @juego.construir_hotel(ip)
            when GestionesInmobiliarias::HIPOTECAR
              @juego.hipotecar(ip)
            when GestionesInmobiliarias::VENDER
              @juego.vender(ip)
            when GestionesInmobiliarias::TERMINAR
              operacion = @juego.siguiente_paso_completado(operacion)
            end
          when OperacionesJuego::SALIR_CARCEL
            respuesta = @vista.salir_carcel
            case @vista.lista_salidas_carcel[respuesta]
            when SalidasCarcel::PAGANDO
              @juego.salir_carcel_pagando
            when SalidasCarcel::TIRANDO
              @juego.salir_carcel_tirando
            end
            @juego.siguiente_paso_completado(operacion)
          end
        end
      end
      puts "Juego terminado. El ránking de los jugadores es: "
      ranking = @juego.ranking
      ranking.each { |jugador| puts jugador.to_s }
    end

  end
end

File: juegoTexto/Controlador.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.GestionesInmobiliarias;
import civitas.OperacionInmobiliaria;
import civitas.OperacionesJuego;
import civitas.Respuestas;
import civitas.SalidasCarcel;

public class Controlador {

  private CivitasJuego juego;
  private VistaTextual vista;

  Controlador(CivitasJuego juego, VistaTextual vista) {
    this.juego = juego;
    this.vista = vista;
  }

  void juega() {
    vista.setCivitasJuego(juego);
    while (!juego.finalDelJuego()) {
      vista.actualizarVista();
      vista.pausa();
      OperacionesJuego sigPaso = juego.siguientePaso();
      vista.mostrarSiguienteOperacion(sigPaso);
      if (sigPaso != OperacionesJuego.PASAR_TURNO) {
        vista.mostrarEventos();
      }
      if (!juego.finalDelJuego()) {
        switch (sigPaso) {
          case COMPRAR:
            if (vista.comprar() == Respuestas.SI) {
              juego.comprar();
            }
            juego.siguientePasoCompletado(sigPaso);
            break;
          case GESTIONAR:
            vista.gestionar();
            GestionesInmobiliarias gestion = GestionesInmobiliarias.values()[vista.getGestion()];
            int ip = vista.getPropiedad();
            OperacionInmobiliaria operacion = new OperacionInmobiliaria(gestion, ip);
            switch (gestion) {
              case VENDER:
                juego.vender(ip);
                break;
              case HIPOTECAR:
                juego.hipotecar(ip);
                break;
              case CANCELAR_HIPOTECA:
                juego.cancelarHipoteca(ip);
                break;
              case CONSTRUIR_CASA:
                juego.construirCasa(ip);
                break;
              case CONSTRUIR_HOTEL:
                juego.construirHotel(ip);
                break;
              case TERMINAR:
                juego.siguientePasoCompletado(sigPaso);
                break;
            }
           break;	
          case SALIR_CARCEL:
            SalidasCarcel modo_salida = vista.salirCarcel();
            switch (modo_salida) {
              case PAGANDO:
                juego.salirCarcelPagando();
                break;
              case TIRANDO:
                juego.salirCarcelTirando();
                break;
            }
            juego.siguientePasoCompletado(sigPaso);
            break;
          default:
            break;
        }
      }
    }
    System.out.println("Juego terminado. El ránking es: ");
    System.out.println(juego.ranking());
  }
} 